### ISSUE TITLE
Support Stingray PBS material in FBX importer

### ISSUE THREAD

**Issue by Mhnix** [assimp/assimp#1504]
Unity 2017.2 gained support for reading Stingray PBS materials imported from FBX files created in Maya.
Now we want to get the same support using assimp.

I have created a rudimentary patch to read most properties from the material, however I am unsure how to correctly expose the PBR specific properties. Is a generic PBR material interface planned (something like AI_MATKEY_METALNESS, AI_MATKEY_ROUGHNESS,.. or similar), or should I add it in a manner similar to the gltf2 PBS properties (AI_MATKEY_GLTF_PBRMETALLICROUGHNESS_METALLIC_FACTOR etc)?

The properties has the basic form "Maya|metallic", "Maya|emissive", etc. Is this too non-standard to be added to assimp mainline?

Attached is a simple fbx file in ascii format (rename to .fbx) with a couple of cubes with different materials whereof two Stingray PBS materials.

[stingray_pbs_fbx2016_ascii.txt](https://github.com/assimp/assimp/files/1399373/stingray_pbs_fbx2016_ascii.txt)
[stingray_pbs.patch.txt](https://github.com/assimp/assimp/files/1399420/stingray_pbs.patch.txt)

**Comment by dhritzkiv** [comment]
+1 for discussing assimp-wide support of PBR materials, including auto-conversion (as best as possible) of traditional lighting <-> PBR.

However, in your case, it probably makes sense to start work similar to gltf2's PBR material properties in the meantime.

Also, for what it's worth, emissive is already supported as `AI_MATKEY_COLOR_EMISSIVE` and `aiTextureType_EMISSIVE`

**Comment by kimkulling** [comment]
+1 for supporting PBR in assimp from my side. From my point of view we have to do this!

**Comment by dhritzkiv** [comment]
I'll open a new Feature request issue!

**Comment by jdolan-zoox** [comment]
Any word on this? Would love to use Assimp but right now FBXSDK is the only way we can get these bindings in.

**Comment by Mhnix** [comment]
I have not worked any more on this.
When looking through the code now it seems some basics for generic PBR support is there (include/assimp/pbrmaterial.h), but the properties seems to be tied to the gltf support.

Over all gltf2 seems to be the way to go for PBR, and I think that will be the way to go for us (the company I work at) if we decide this issue is critical enough (Maya has since gained some gltf2 exporters in the form of Maya2glTF and babylon.js).

**Comment by jdolan-zoox** [comment]
We want to directly import assets in the format artists store them, and GLTF just isn't really usable as an interchange format. Its great to see Assimp already has *some* effort put towards PBR though!

I did find that https://github.com/facebookincubator/FBX2glTF seems to do a fairly decent job at taking FBX files with Stingray materials and generating GLTF with plausible PBR materials. Granted, I don't want to complicate our pipeline by adding more intermediate steps, but FBX2glTF I hope is a reasonable reference (https://github.com/facebookincubator/FBX2glTF/blob/master/src/Fbx2Raw.cpp#L127) for how that mapping could happen.

Anyways, if I had more free time on my hands I would work out how to add support directly to Assimp and push a PR, but right now its looking more straight forward for us to just use FBXSDK for FBX importing. Which of course isn't ideal as FBXSDK is a pain in the ass in more ways than one :p

**Comment by fire** [comment]
Did you investigate arnold 5 conversion? There is a new fbx material property setup.

**Comment by fire** [comment]
Here are the Arnold and Stingray defines I've been using.

```cpp
#define AI_MATKEY_FBX_MAYA_BASE_COLOR_FACTOR "$raw.Maya|baseColor", 0, 0
#define AI_MATKEY_FBX_MAYA_METALNESS_FACTOR "$raw.Maya|metalness", 0, 0
#define AI_MATKEY_FBX_MAYA_DIFFUSE_ROUGHNESS_FACTOR "$raw.Maya|diffuseRoughness", 0, 0
#define AI_MATKEY_FBX_EMISSIVE_FACTOR "$raw.Maya|emissionColor", 0, 0
#define AI_MATKEY_FBX_EMISSIVE_INTENSITY_FACTOR aiTextureType_UNKNOWN, 0

#define AI_MATKEY_FBX_MAYA_METALNESS_TEXTURE "$raw.Maya|metalness|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_METALNESS_UV_XFORM "$raw.Maya|metalness|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_DIFFUSE_ROUGHNESS_TEXTURE "$raw.Maya|diffuseRoughness|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_DIFFUSE_ROUGHNESS_UV_XFORM "$raw.Maya|diffuseRoughness|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_BASE_COLOR_TEXTURE "$raw.Maya|baseColor|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_BASE_COLOR_UV_XFORM "$raw.Maya|baseColor|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_NORMAL_TEXTURE "$raw.Maya|normalCamera|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_NORMAL_UV_XFORM "$raw.Maya|normalCamera|uvtrafo", aiTextureType_UNKNOWN, 0

#define AI_MATKEY_FBX_MAYA_STINGRAY_DISPLACEMENT_SCALING_FACTOR "$raw.Maya|displacementscaling", 0, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_BASE_COLOR_FACTOR "$raw.Maya|base_color", 0, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_EMISSIVE_FACTOR "$raw.Maya|emissive", 0, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_METALLIC_FACTOR "$raw.Maya|metallic", 0, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_ROUGHNESS_FACTOR "$raw.Maya|roughness", 0, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_EMISSIVE_INTENSITY_FACTOR "$raw.Maya|emissive_intensity", 0, 0

#define AI_MATKEY_FBX_MAYA_STINGRAY_NORMAL_TEXTURE "$raw.Maya|TEX_normal_map|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_NORMAL_UV_XFORM "$raw.Maya|TEX_normal_map|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_COLOR_TEXTURE "$raw.Maya|TEX_color_map|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_COLOR_UV_XFORM "$raw.Maya|TEX_color_map|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_METALLIC_TEXTURE "$raw.Maya|TEX_metallic_map|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_METALLIC_UV_XFORM "$raw.Maya|TEX_metallic_map|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_ROUGHNESS_TEXTURE "$raw.Maya|TEX_roughness_map|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_ROUGHNESS_UV_XFORM "$raw.Maya|TEX_roughness_map|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_EMISSIVE_TEXTURE "$raw.Maya|TEX_emissive_map|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_EMISSIVE_UV_XFORM "$raw.Maya|TEX_emissive_map|uvtrafo", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_AO_TEXTURE "$raw.Maya|TEX_ao_map|file", aiTextureType_UNKNOWN, 0
#define AI_MATKEY_FBX_MAYA_STINGRAY_AO_UV_XFORM "$raw.Maya|TEX_ao_map|uvtrafo", aiTextureType_UNKNOWN, 0
```

This list is missing |uvwsrc.

**Comment by b1skit** [comment]
+1 for Maya PBR shader support!

**Comment by RevoluPowered** [comment]
PBR support has been PRd https://github.com/assimp/assimp/pull/2640

Will follow up with PBS in a future PR.

**Comment by RevoluPowered** [comment]
Has anyone got a PBS model I can test with as I believe I have implemented it but I need some model to test on?

**Comment by RevoluPowered** [comment]
Stingray added :)
https://github.com/assimp/assimp/pull/2640

**Comment by arodic** [comment]
Should this issue be closed considering #2640 was merged?

**Comment by kimkulling** [comment]
Right, just forget to do it. Sorry!
